webapp/tests/karma/ts/services/task-notifications.service.spec.ts

```diff
@@ -0,0 +1,241 @@
+    expect(notifications).to.be.an('array').that.has.lengthOf(2);
+    expect(notifications[0]._id).to.equal('task2');
+    expect(notifications[1]._id).to.equal('task1');
+    expect(rulesEngine.fetchTaskDocsForAllContacts.callCount).to.equal(1);
+    expect(translateService.instant.callCount).to.equal(2);
+    expect(consoleErrorMock.callCount).to.equal(0);
+
+    notifications.forEach((notification) => {
+      const task = getTask(notification._id);
+      expect(notification).to.be.an('object');
+      expect(notification).to.eql({
+        _id: task._id,
+        readyAt: task.stateHistory[0].timestamp,
+        title: task.emission.title,
+        contentText: 'android.notification.tasks.contentText',
+        dueDate: task.emission.dueDate
+      });
+    });
+  });
+
+  it('should not return same notifications twice', async () => {
+    const notifications = await service.fetchNotifications();
+    expect(notifications).to.be.an('array').that.has.lengthOf(2);
+    const secondCallNotifications = await service.fetchNotifications();
+    expect(secondCallNotifications).to.be.an('array').that.has.lengthOf(0);
+  });
+
+  it('should return notification if new task is generated', async () => {
+    const notifications = await service.fetchNotifications();
+    expect(notifications).to.be.an('array').that.has.lengthOf(2);
+    const newTask = {
+      _id: 'task3',
+      state: 'Ready',
+      emission: {
+        title: 'Task 3',
+        contact: { name: 'Owl Phil3', _id: 'contact3' },
+        dueDate: moment().format('YYYY-MM-DD')
+      },
+      stateHistory: [
+        {
+          state: 'Ready',
+          timestamp: moment().add(120, 'milliseconds').valueOf(),
+        }
+      ]
+    };
+    rulesEngine.fetchTaskDocsForAllContacts.resolves([...tasks, newTask]);
+    const updatedNotifications = await service.fetchNotifications();
+    expect(updatedNotifications).to.be.an('array').that.has.lengthOf(1);
+    expect(updatedNotifications[0]._id).to.equal(newTask._id);
+  });
+
+  it('should return task notification due today on a new day', async () => {
+    const taskDueTomorrow = {
+      _id: 'task_tomorrow',
+      state: 'Ready',
+      emission: {
+        title: 'Task 3',
+        contact: { name: 'Future Owl', _id: 'contact_future' },
+        dueDate: moment().add(1, 'day').format('YYYY-MM-DD')
+      },
+      stateHistory: [
+        {
+          state: 'Ready',
+          timestamp: moment().subtract(1, 'day').valueOf(),
+        }
+      ]
+    };
+
+    rulesEngine.fetchTaskDocsForAllContacts.resolves([...tasks, taskDueTomorrow]);
+    const clock = sinon.useFakeTimers(moment().valueOf());
+    const notifications = await service.fetchNotifications();
+    expect(notifications).to.be.an('array').that.has.lengthOf(2);
+
+    clock.tick(DAY_IN_MS);
+
+    const updatedNotifications = await service.fetchNotifications();
+    expect(updatedNotifications).to.be.an('array').that.has.lengthOf(3);
+    clock.restore();
```

Comment:
The issue with restoring the clock at the end of the test is that, if the test fails in an assertion, the clock is not restored. 
My preferred approach is to declare `clock` as a higher scope variable, and do a `clock?.restore()` in the `afterEach` hook